### PR TITLE
[SPARK-29015][SQL][test-hadoop3.2]Reset class loader after initializing SessionState for built-in Hive 2.3

### PR DESCRIPTION
**Pull request by AngersZhuuuu**
### What changes were proposed in this pull request?

Hive 2.3 will set a new UDFClassLoader to hiveConf.classLoader when initializing SessionState since HIVE-11878,  and 
1. ADDJarCommand will add jars to clientLoader.classLoader.
2. --jar passed jar will be added to clientLoader.classLoader
3.  jar passed by hive conf  `hive.aux.jars`  [SPARK-28954](https://github.com/apache/spark/pull/25653) [SPARK-28840](https://github.com/apache/spark/pull/25542) will be added to clientLoader.classLoader too

For these  reason we cannot load the jars added by ADDJarCommand because of class loader got changed. We reset it to clientLoader.ClassLoader here.

### Why are the changes needed?
support for jdk11

### Does this PR introduce any user-facing change?
NO

### How was this patch tested?
UT
```
export JAVA_HOME=/usr/lib/jdk-11.0.3
export PATH=$JAVA_HOME/bin:$PATH

build/sbt -Phive-thriftserver -Phadoop-3.2

hive/test-only *HiveSparkSubmitSuite -- -z "SPARK-8368: includes jars passed in through --jars"
hive-thriftserver/test-only *HiveThriftBinaryServerSuite -- -z "test add jar"
```

### CODE REVIEW

**Comment by AngersZhuuuu** [comment]
gentle ping @dongjoon-hyun  can you help yo trigger again for jdk8

File: sql/hive/src/main/scala/org/apache/spark/sql/hive/execution/HiveTableScanExec.scala [hunk 1]
```diff
@@ -120,7 +121,16 @@ case class HiveTableScanExec(
 
     HiveShim.appendReadColumns(hiveConf, neededColumnIDs, output.map(_.name))
 
-    val deserializer = tableDesc.getDeserializerClass.getConstructor().newInstance()
+    val currentState = SessionState.get()
+    val deserializer = if (currentState != null) {
+      val originClassLoader = currentState.getConf.getClassLoader
+      currentState.getConf.setClassLoader(sparkSession.sharedState.jarClassLoader)
```
Comment (wangyum) [inline sql/hive/src/main/scala/org/apache/spark/sql/hive/execution/HiveTableScanExec.scala]: Could we set the  correct class loader when we constructing `SessionState`? Maybe it's here:
https://github.com/apache/spark/blob/128ea37bda3dbc1d6ba2af762c7453ab7605d430/sql/hive/src/main/scala/org/apache/spark/sql/hive/client/HiveClientImpl.scala#L189-L193

Comment (wangyum) [reply sql/hive/src/main/scala/org/apache/spark/sql/hive/execution/HiveTableScanExec.scala]: We fixed `CliSessionState` class loader issue when constructing it by [SPARK-28840](https://issues.apache.org/jira/browse/SPARK-28840), so could we fix `SessionState` class loader issue when constructing `SessionState`?


Hive related code:
https://github.com/apache/hive/blob/rel/release-2.3.6/ql/src/java/org/apache/hadoop/hive/ql/session/SessionState.java#L399

Comment (AngersZhuuuu) [reply sql/hive/src/main/scala/org/apache/spark/sql/hive/execution/HiveTableScanExec.scala]: > We fixed `CliSessionState` class loader issue when constructing it by [SPARK-28840](https://issues.apache.org/jira/browse/SPARK-28840), so could we fix `SessionState` class loader issue when constructing `SessionState`?

It's better to solve this when construct HiveClientImpl's SessionState.
And we can handle it with same method like [SPARK-28840](https://issues.apache.org/jira/browse/SPARK-28840)

Comment (wangyum) [reply sql/hive/src/main/scala/org/apache/spark/sql/hive/execution/HiveTableScanExec.scala]: Thank you @AngersZhuuuu

Comment (wangyum) [reply sql/hive/src/main/scala/org/apache/spark/sql/hive/execution/HiveTableScanExec.scala]: Another question is why this issue only occurs in JDK 11?

Comment (AngersZhuuuu) [reply sql/hive/src/main/scala/org/apache/spark/sql/hive/execution/HiveTableScanExec.scala]: > Another question is why this issue only occurs in JDK 11?

Still work on this to find root cause. Will add to desc when find it.

File: sql/hive/src/main/scala/org/apache/spark/sql/hive/client/HiveClientImpl.scala [hunk 2]
```diff
@@ -190,6 +190,7 @@ private[hive] class HiveClientImpl(
     if (clientLoader.cachedHive != null) {
       Hive.set(clientLoader.cachedHive.asInstanceOf[Hive])
     }
+    state.getConf.setClassLoader(initClassLoader)
```
Comment (wangyum) [inline sql/hive/src/main/scala/org/apache/spark/sql/hive/client/HiveClientImpl.scala]: We need to add comments to explain this change.

Comment (wangyum) [reply sql/hive/src/main/scala/org/apache/spark/sql/hive/client/HiveClientImpl.scala]: Could we just set the `ClassLoader` for Hive 2.3?
```scala
if (HiveUtils.isHive23) {
  state.getConf.setClassLoader(clientLoader.classLoader)
}
```

Comment (wangyum) [reply sql/hive/src/main/scala/org/apache/spark/sql/hive/client/HiveClientImpl.scala]: Could you help fix [this incorrect comment](https://github.com/apache/spark/pull/25775/files#diff-6fd847124f8eae45ba2de1cf7d6296feR164): `initialConf.setClassLoader(initClassLoader)` -> `hiveConf.setClassLoader(initClassLoader)`

**Comment by wangyum** [comment]
This PR tested by [HiveThriftBinaryServerSuite#test add jar](https://github.com/apache/spark/blob/962e330955581aea032ff336a12f23374c39e672/sql/hive-thriftserver/src/test/scala/org/apache/spark/sql/hive/thriftserver/HiveThriftServer2Suites.scala#L491-L535)?

File: sql/hive/src/main/scala/org/apache/spark/sql/hive/client/HiveClientImpl.scala (hunk 2 above)
Comment (AngersZhuuuu) [reply sql/hive/src/main/scala/org/apache/spark/sql/hive/client/HiveClientImpl.scala]: > We need to add comments to explain this change.

Will do,  but 
```
if (HiveUtils.isHive23) {
  state.getConf.setClassLoader(initClassLoader)
}
```
can't work with  version higher then 2.3

Comment (AngersZhuuuu) [reply sql/hive/src/main/scala/org/apache/spark/sql/hive/client/HiveClientImpl.scala]: > Could you help fix [this incorrect comment](https://github.com/apache/spark/pull/25775/files#diff-6fd847124f8eae45ba2de1cf7d6296feR164): `initialConf.setClassLoader(initClassLoader)` -> `hiveConf.setClassLoader(initClassLoader)`

Done

**Comment by wangyum** [comment]
@AngersZhuuuu Could you add these to **How was this patch tested?**:
```shell
export JAVA_HOME=/usr/lib/jdk-11.0.3
export PATH=$JAVA_HOME/bin:$PATH

build/sbt -Phive-thriftserver -Phadoop-3.2

hive/test-only *HiveSparkSubmitSuite -- -z "SPARK-8368: includes jars passed in through --jars"
hive-thriftserver/test-only *HiveThriftBinaryServerSuite -- -z "test add jar"
```

**Comment by srowen** [comment]
@wangyum per your comment at https://github.com/apache/spark/pull/25690 let's merge this first. It looks like it has passed a few times, but I see a number of failures too; since you have been following this more closely, they were due to previous versions of the change? the comment and change sound plausible.

You an also edit the description directly if it's faster. I will make your last change to the description.

**Comment by wangyum** [comment]
I see. Thank you @srowen

**Comment by dongjoon-hyun** [comment]
Could we revise the PR title more specifically, @AngersZhuuuu and @wangyum ?

File: sql/hive/src/main/scala/org/apache/spark/sql/hive/client/HiveClientImpl.scala [hunk 3]
```diff
@@ -190,6 +190,13 @@ private[hive] class HiveClientImpl(
...
       Hive.set(clientLoader.cachedHive.asInstanceOf[Hive])
     }
+    // Hive 2.3 will set UDFClassLoader to hiveConf when initializing SessionState
+    // since HIVE-11878, and ADDJarCommand will add jars to clientLoader.classLoader.
+    // For this reason we cannot load the jars added by ADDJarCommand because of class loader
+    // got changed. We reset it to clientLoader.ClassLoader here.
+    if (HiveUtils.isHive23) {
+      state.getConf.setClassLoader(clientLoader.classLoader)
+    }
     SessionState.start(state)
```
Comment (wangyum) [inline sql/hive/src/main/scala/org/apache/spark/sql/hive/client/HiveClientImpl.scala]: @AngersZhuuuu I think we need to add another test case simily to [`HiveSparkSubmitSuite#SPARK-8368: includes jars passed in through --jars`](https://github.com/apache/spark/blob/016e1b491c9099063df8e2e76bc58d64fc369490/sql/hive/src/test/scala/org/apache/spark/sql/hive/HiveSparkSubmitSuite.scala#L110-L127) but with `spark.sql.hive.metastore.version=3.1` and `spark.sql.hive.metastore.jars=maven`.

Comment (wangyum) [reply sql/hive/src/main/scala/org/apache/spark/sql/hive/client/HiveClientImpl.scala]: Sorry. It's another bug: https://issues.apache.org/jira/browse/SPARK-29254

File: sql/hive/src/main/scala/org/apache/spark/sql/hive/client/HiveClientImpl.scala [hunk 4]
```diff
@@ -190,6 +190,13 @@ private[hive] class HiveClientImpl(
     if (clientLoader.cachedHive != null) {
       Hive.set(clientLoader.cachedHive.asInstanceOf[Hive])
     }
+    // Hive 2.3 will set UDFClassLoader to hiveConf when initializing SessionState
+    // since HIVE-11878, and ADDJarCommand will add jars to clientLoader.classLoader.
+    // For this reason we cannot load the jars added by ADDJarCommand because of class loader
+    // got changed. We reset it to clientLoader.ClassLoader here.
+    if (HiveUtils.isHive23) {
+      state.getConf.setClassLoader(clientLoader.classLoader)
+    }
```
Comment (wangyum) [inline sql/hive/src/main/scala/org/apache/spark/sql/hive/client/HiveClientImpl.scala]: @gatorsmile @srowen  @HyukjinKwon @dongjoon-hyun @juliuszsompolski Do you have more comments?

Comment (juliuszsompolski) [reply sql/hive/src/main/scala/org/apache/spark/sql/hive/client/HiveClientImpl.scala]: No opinion from me, I know too little context of how these classloaders should be used to review.

File: sql/hive/src/main/scala/org/apache/spark/sql/hive/client/HiveClientImpl.scala (hunk 3 above)
Comment (wangyum) [reply sql/hive/src/main/scala/org/apache/spark/sql/hive/client/HiveClientImpl.scala]: SPARK-29254 is a different issue.

**Review by srowen — APPROVED** [review]

**Comment by srowen** [comment]
Merged to master

File: sql/hive/src/main/scala/org/apache/spark/sql/hive/client/HiveClientImpl.scala [hunk 5]
```diff
@@ -190,6 +190,13 @@ private[hive] class HiveClientImpl(
     if (clientLoader.cachedHive != null) {
       Hive.set(clientLoader.cachedHive.asInstanceOf[Hive])
     }
+    // Hive 2.3 will set UDFClassLoader to hiveConf when initializing SessionState
+    // since HIVE-11878, and ADDJarCommand will add jars to clientLoader.classLoader.
+    // For this reason we cannot load the jars added by ADDJarCommand because of class loader
+    // got changed. We reset it to clientLoader.ClassLoader here.
+    if (HiveUtils.isHive23) {
```
Comment (yaooqinn) [inline sql/hive/src/main/scala/org/apache/spark/sql/hive/client/HiveClientImpl.scala:197]: Why not set the classLoader blindly to `clientLoader.classLoader`? HIVE-11878 got merged into hive 1.3.0 and 2.0.0? This issue can still exist in the production environment when we try to use non-builtin hive libraries.

cc @ulysses-you @cloud-fan @wangyum @AngersZhuuuu

**Pull request closed without merging**